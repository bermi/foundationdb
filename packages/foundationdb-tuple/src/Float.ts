/**
 * Wrapper object for float values
 */
export class Float {
    /**
     * Actual value of a float tuple
     */
    readonly value: number;

    constructor(value: number) {
        this.value = value;
        Object.freeze(this);
    }
}