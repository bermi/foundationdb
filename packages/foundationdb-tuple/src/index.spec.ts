import { Float } from './Float';
import { unpack, TupleItem, pack } from './index';

function testConsistency(source: TupleItem[]) {
    let res = pack(source);
    let decoded = unpack(res);
    let encoded = pack(decoded);
    expect(encoded.equals(res)).toBe(true);
}

function testConsistencyBinary(source: Buffer) {
    let decoded = unpack(source);
    let encoded = pack(decoded);
    expect(encoded.equals(source)).toBe(true);
}

function testConsistencyBoth(source: TupleItem[], sourceBin: Buffer) {
    let res = pack(source);
    let decoded = unpack(res);
    let encoded = pack(decoded);
    expect(encoded.equals(res)).toBe(true);
    expect(sourceBin.equals(encoded)).toBe(true);
}

function snapshotTest(name: string, tuple: TupleItem[]) {
    it('should encode ' + name, () => {
        let res = pack(tuple);
        let unpacked = unpack(res);
        expect(unpacked.length).toBe(tuple.length);
        for (let i = 0; i < tuple.length; i++) {
            if (tuple[i] instanceof Float) {
                expect(unpacked[i] instanceof Float).toBe(true);
                expect((unpacked[i] as Float).value).toBe((tuple[i] as Float).value);
            } else if (Buffer.isBuffer(tuple[i])) {
                expect(Buffer.isBuffer(unpacked[i])).toBe(true);
                expect((unpacked[i] as Buffer).equals(tuple[i] as Buffer)).toBe(true);
            } else {
                expect(unpacked[i] === tuple[i]).toBe(true);
            }
        }
        expect(res.toString('hex')).toMatchSnapshot();
    });
}

describe('tuple', () => {
    it('should preserve -0', () => {
        testConsistencyBoth([new Float(-0)], Buffer.from('217fffffffffffffff', 'hex'));
    });
    it('should preserve integers', () => {
        testConsistencyBinary(Buffer.from('218000000000000000', 'hex'));
    });
    it('should preserve PI', () => {
        testConsistency([new Float(Math.PI)]);
    });
    it('should throw error on NaN', () => {
        expect(() => unpack(Buffer.from('21fff8000000000000', 'hex'))).toThrowError();
        expect(() => unpack(Buffer.from('210007ffffffffffff', 'hex'))).toThrowError();
    });
    snapshotTest('mixed data', ['user', 1, 'score', new Float(0.4)]);
    snapshotTest('integers', [0, 1, -1, 2, 10, 10000, Math.pow(2, 50), -0, -10000, -45553]);
    snapshotTest('strings', ['', 'hello!', 'ðŸ¤£', '??', '\x00', '\x00\xff\x00']);
    snapshotTest('floats', [new Float(Math.PI), new Float(1), new Float(-0), new Float(0), new Float(Math.pow(2, 100))]);
});