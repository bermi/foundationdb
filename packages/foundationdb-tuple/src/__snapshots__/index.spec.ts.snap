// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tuple should encode floats 1`] = `"21c00921fb54442d1821bff0000000000000217fffffffffffffff21800000000000000021c630000000000000"`;

exports[`tuple should encode integers 1`] = `"14150113fe1502150a1627101b040000000000001412d8ef124e0e"`;

exports[`tuple should encode mixed data 1`] = `"02757365720015010273636f72650021bfd999999999999a"`;

exports[`tuple should encode strings 1`] = `"02000268656c6c6f210002f09fa4a300023f3f000200ff000200ffc3bf00ff00"`;
